### PR TITLE
Support for zooming from scripting

### PR DESCRIPTION
**Pull request by Semphriss**
## Zooming now available to level designers!

zooming

If you have a portion of you level that is bigger or smaller, you can now use zooming to increase the effect on the player :)

Accessible from `Camera.set_scale(scale)`, `Camera.scale(scale, time)` and `Camera.ease_scale(scale, time, easing)` where:
- **scale** is a decimal number which is the zoom multiplier. 1 is default zoom, The gif uses 0.75. A smaller number = a larger portion of the level visible to the screen.
- **time** is a decimal number that represent the time it will take to reach the scale.
- **easing** is a string with the name of the easing you want to use, following the format "Ease`[Mode][Direction]`" where "Mode" is the easing mode (Sine, Exponential, Back, Bounce, etc.) and Direction is In, Out, or InOut. For a ful llist of possibilites, see: https://easings.net/ (note that the name are not exactly the name, the order is different). Examples: "EaseSineInOut", "EaseBounceOut", "EaseBackIn"

### CODE REVIEW

**Comment by Rusty-Box** [comment]
Is this linked to the magnification setting? And if so, can you also return the zoom back to the original setting?

**Comment by Semphriss** [comment]
This works in a similar way but is completely independant from the magnification setting. The scale is tied to the camera object, and is applied only to the sector the camera is linked to, similarly to the camera translation.

**Comment by Rusty-Box** [comment]
So, a 200% zoom when magnification is 100 is different from a 200% zoom when magnification is 160?

**Comment by Semphriss** [comment]
Yes, indeed. A magnification of 200 with a zoom of 200% would result in an overall zoom of 400%.

**Comment by Rusty-Box** [comment]
Okay, but doesn't this provide a problem for players with different different settings. For example, if a cutscene zoomes in (based on a magnification of 100) important things may get lost if the same cutscene is viewed by players having a magnification of 200?

**Comment by Semphriss** [comment]
This problem isn't introduced by this feature, though. Players on smaller/bigger resolutions have the same problem. The difference between what players on 640x480 see and those on 1920x1080 see will be the same at 100% zoom or 500% zoom.

This is a level design thing, and since the feature is for level designers, it will be their job to make sure what they want to show is clear and visible on small resolutions. This is true without this feature, and it will be the same with it :)

File: src/object/camera.hpp [hunk 1]
```diff
@@ -97,12 +97,22 @@ class Camera final : public GameObject,
   Vector get_center() const;
 
   void set_mode(Mode mode_) { m_mode = mode_; }
+
+  /** get the exact scale at this exact moment */
+  float get_current_scale() { return m_scale; }
```
Comment (tobbi) [inline src/object/camera.hpp:102]: This can be marked `const`

File: src/object/camera.hpp [hunk 2]
```diff
@@ -97,12 +97,22 @@ class Camera final : public GameObject,
...
+
+  /** get the scale towards which the camera is moving */
+  float get_target_scale() { return m_scale_target; }
```
Comment (tobbi) [inline src/object/camera.hpp:105]: This too.

File: src/video/canvas.cpp [hunk 3]
```diff
@@ -127,7 +127,9 @@ Canvas::draw_surface(const SurfacePtr& surface,
   request->blend = blend;
 
   request->srcrects.emplace_back(Rectf(surface->get_region()));
-  request->dstrects.emplace_back(Rectf(apply_translate(position), Size(surface->get_width(), surface->get_height())));
+  request->dstrects.emplace_back(Rectf(apply_translate(position)*scale(),
```
Comment (tobbi) [inline src/video/canvas.cpp:130]: Add spaces around the `*`

Comment (Semphriss) [reply src/video/canvas.cpp:130]: I've rarely seen so many consecutive closing parentheses.

**Pull request merged**